### DISCUSSION TITLE
gcd time limit

### DISCUSSION THREAD

**Discussion by DesmondZhong** [sicmutils/sicmutils#332] [Show and tell]
Hello everyone!

This is my first post here! I was going through the examples in "Functional Differential Geometry" and trying to validate Eqn 4.34 on P49, when I encountered an unexpected behavior. I also found a similar block of code in the [test](https://github.com/sicmutils/sicmutils/blob/v0.16.0/test/sicmutils/fdg/ch4_test.cljc#L207), so I guess if the release (0.16.0) passed all the test then I shouldn't encounter this. Here's my code

```clojure
;; a few definitions
(def polar-vector-basis (coordinate-system->vector-basis R2-polar))
(def polar-dual-basis (coordinate-system->oneform-basis R2-polar))
(def f-rect (literal-manifold-function 'f-rect R2-rect))
(def e0 (* (coordinate-system->vector-basis R2-rect)
            (up (literal-manifold-function 'e0x R2-rect)
                (literal-manifold-function 'e0y R2-rect))))
(def e1 (* (coordinate-system->vector-basis R2-rect)
            (up (literal-manifold-function 'e1x R2-rect)
                (literal-manifold-function 'e1y R2-rect))))
;; here is main block of code
(zero?
 (simplify
  ((- ((commutator e0 e1) f-rect)
      (* (- (e0 (polar-dual-basis e1))
            (e1 (polar-dual-basis e0)))
         (polar-vector-basis f-rect)))
   R2-rect-point)))
;; this should return true but it returns false
```
I realized in the [test code](https://github.com/sicmutils/sicmutils/blob/v0.16.0/test/sicmutils/fdg/ch4_test.cljc#L219), there are two lines of code
```clojure
binding [pg/*poly-gcd-time-limit* #?(:clj  [1 :seconds]
                                     :cljs [4 :seconds])]
```
which seems to increase the time limit for gcd. I guess maybe the default value is not able to let the simplification process finishf, so I modified my code to be
```clojure
(binding [pg/*poly-gcd-time-limit* #?(:clj  [1 :seconds]
                                      :cljs [4 :seconds])]
 (simplify
  ((- ((commutator e0 e1) f-rect)
      (* (- (e0 (polar-dual-basis e1))
            (e1 (polar-dual-basis e0)))
         (polar-vector-basis f-rect)))
   R2-rect-point)))
;; still returns a complicated expression instead of 0
```
Did I miss anything or how can I properly set the time limit? I appreciate any advice on this. Thanks!

P.S. To rule out the possibility of typos or stupid errors, I plug in random functions and numbers into the expression as follows
```clojure
;; define specific functions and specific points
(defn f-R2->R
    [up-vector]
    (* (nth up-vector 0)
       (nth up-vector 1)))
(def a-f-rect
    (compose f-R2->R (chart R2-rect)))
(def a-R2-rect-point ((point R2-rect) (up 3 2)))
(let-coordinates
 [[x y] R2-rect]
 (def a-e0
     (+ (* x d:dx)
        (* y d:dy)))
 (def a-e1
     (+ (* y y d:dx)
        (* x x d:dy))))
;; plug in a specific function and a specific point
(simplify
 ((- ((commutator a-e0 a-e1) a-f-rect)
     (* (- (a-e0 (polar-dual-basis a-e1))
           (a-e1 (polar-dual-basis a-e0)))
        (polar-vector-basis a-f-rect)))
  a-R2-rect-point))
;; this returns 0.0, even after I change the function and point
```

**Comment by sritchie** [comment]
Hi @DesmondZhong, and welcome to our tiny-but-mighty discussion forum :)

You startled me because there IS a bug in the library associated with the GCD time limit, but I don't think you're hitting it here. (If you were you'd see a warning say "GCD time limit reached!" or something like that.) First, your issue, then I'll say more...

The only difference between the code you posted and the code in the tests is the multiplication order of the terms in `e0` and `e1`, and this matters because operator multiplication (function composition) isn't associative. Switch your definitions in for these and you'll get 0:

```clojure
(def e0
  (* (up (literal-manifold-function 'e0x R2-rect)
         (literal-manifold-function 'e0y R2-rect))
     (coordinate-system->vector-basis R2-rect)))

(def e1
  (* (up (literal-manifold-function 'e1x R2-rect)
         (literal-manifold-function 'e1y R2-rect))
     (coordinate-system->vector-basis R2-rect)))
```

And about this:

> P.S. To rule out the possibility of typos or stupid errors, I plug in random functions and numbers into the expression as follows

The relationship is true of any manifold function or point, which is why you're getting `0` for specific inputs.

@DesmondZhong let me know if this makes sense... and more importantly let me know if it fixes your issue!

### GCD

Now, the GCD issue, If you're curious... the problem with the current implementation is that if you DO hit that timeout, you'll see a warning, but the simplifier will also cache that expression, and will NOT try to simplify it the next time you pass it in. This is broken because you can't adjust the timeout and try again.

That's wrong, and this is a good nudge to fix that, even though it didn't apply directly here!

**Reply by sritchie** [reply]
@DesmondZhong you might also like the new docs for `let-coordinates`, which show what exactly that macro is doing, and what bindings are available inside:

https://cljdoc.org/d/sicmutils/sicmutils/0.17.0/api/sicmutils.calculus.coordinate#let-coordinates

**Reply by DesmondZhong** [reply]
Thanks Sam! This indeed solves the issue! 

I further investigate this a little bit. Let's look at the code where I plug in a specific function and point. If I change the definition `a-e0`, `a-e1` to 
```clojure
 (def a-e0
     (+ (* d:dx x)
        (* d:dy y)))
 (def a-e1
     (+ (* d:dx y y)
        (* d:dy x x)))
```
the expression gives 0, but this is actually due to numerical error. Change the point to, for example, `(def a-R2-rect-point ((point R2-rect) (up 33 22)))` would results in a nonzero value of the expression (`-2.1827872842550278E-11` in this case.) 
This shows that the expression is not evaluated to be zero at every point, which is consistent with "simplifying the symbolic expression doesn't return 0", as we have seen. This numerical error is probably why I didn't find true issue in the first place. 

As you said, 
>operator multiplication (function composition) isn't commutative

I think the following code snippets might be easier for people to understand what's happening here.

```clojure
(let-coordinates
 [[x y] R2-rect]
 (simplify
  (((* x d:dx) x) R2-rect-point)))
;; returns "x0"
(let-coordinates
 [[x y] R2-rect]
 (simplify
  (((* d:dx x) x) R2-rect-point)))
;; returns "(* 2 x0)"
(let-coordinates
 [[x y] R2-rect]
 (simplify
  ((d:dx (* x x)) R2-rect-point)))
;; returns "(* 2 x0)"
```
In general, the vector field `(* d:dx g-rect)` applying to a manifold function `f-rect` produces a manifold function which is the same as the vector field `d:dx` applying to a manifold function `(* g-rect f-rect)` (and also `(* f-rect g-rect)` since multiplication of manifold functions are commutative!.) This is shown in the following code snippet.
```clojure
(def f-rect (literal-manifold-function 'f-rect R2-rect))
(def g-rect (literal-manifold-function 'g-rect R2-rect))
(let-coordinates
 [[x y] R2-rect]
 (simplify
  (- (((* d:dx g-rect) f-rect) R2-rect-point)
     ((d:dx (* f-rect g-rect)) R2-rect-point))))
;; returns "0"
```

Thanks again Sam! I think I've made similar mistakes before and I'll go through my notebooks and correct them. 

The new docs for `let-coordinates` looks awesome! Although I've got an vague idea of what bindings are available as I studied FDG, it is nice to see it formally documented. :-)

**Reply by sritchie** [reply]
Hey @DesmondZhong , I'll read more closely post dinner but just wanted to point out a mistake of mine that I think you've seen: I meant that operator multiplication isn't _commutative_, but is in fact associative!

Excellent stuff, will read shortly and edit this comment!

**Reply by DesmondZhong** [reply]
Hi Sam, I've edited the quote in my comment. :-)

I'm thinking about changing the title of this thread to something like "multiplication between manifold function and basis vector field is not commutative." The current one is misleading for people. What do you think?
